File: boxsvg/generators/mailer.py
"""Mailer box (roll-end tuck-top) dieline generator.

Layout (unfolded, bottom-up):

    Bottom dust flaps
    ─── score ───
    Bottom panel    (length × width)
    ─── score ───
    Front panel     (length × height)
    ─── score ───
    Top panel       (length × width)
    ─── score ───
    Top tuck flap   (length × tuck_depth)

Side flaps extend from the left and right edges of each panel.
"""

from __future__ import annotations

from boxsvg.models import BoxRequest, Dieline, Line


def _hline(x1: float, x2: float, y: float, kind: str) -> Line:
    return Line(x1=x1, y1=y, x2=x2, y2=y, kind=kind)


def _vline(x: float, y1: float, y2: float, kind: str) -> Line:
    return Line(x1=x, y1=y1, x2=x, y2=y2, kind=kind)


def generate_mailer_dieline(request: BoxRequest) -> Dieline:
    L = request.length  # along the horizontal center
    W = request.width
    H = request.height
    t = request.effective_thickness()
    kerf = request.effective_kerf()
    fa = request.fold_allowance()

    # Flap dimensions — reduced by thickness to prevent overlap when folded
    flap_w = W / 2 - t
    tuck_depth = W * 0.75   # tuck flap depth
    dust_flap_h = W * 0.5   # bottom dust flaps

    # Panel heights with fold allowance compensation
    bottom_panel_h = W                # base panel, no compensation needed
    front_panel_h = H + fa            # wraps around bottom panel material
    top_panel_h = W + fa              # wraps over front panel material

    # Total layout dimensions
    # Horizontal: flap_w | L | flap_w
    # Vertical (bottom to top): dust_flap_h | bottom | front | top | tuck
    total_w = flap_w + L + flap_w
    total_h = dust_flap_h + bottom_panel_h + front_panel_h + top_panel_h + tuck_depth

    # Key x coordinates
    x_left_flap = 0.0
    x_body_left = flap_w
    x_body_right = flap_w + L
    x_right_flap = total_w

    # Key y coordinates (from bottom)
    y_bottom = total_h
    y_dust_top = total_h - dust_flap_h
    y_bottom_panel_top = y_dust_top - bottom_panel_h
    y_front_panel_top = y_bottom_panel_top - front_panel_h
    y_top_panel_top = y_front_panel_top - top_panel_h
    y_tuck_top = 0.0

    lines: list[Line] = []

    # === BOTTOM DUST FLAPS ===
    # Left dust flap
    lines.append(_hline(x_left_flap, x_body_left, y_dust_top, "cut"))        # top edge
    lines.append(_vline(x_left_flap, y_dust_top, y_bottom, "cut"))            # left edge
    lines.append(_hline(x_left_flap, x_body_left, y_bottom, "cut"))           # bottom edge

    # Right dust flap
    lines.append(_hline(x_body_right, x_right_flap, y_dust_top, "cut"))      # top edge
    lines.append(_vline(x_right_flap, y_dust_top, y_bottom, "cut"))           # right edge
    lines.append(_hline(x_body_right, x_right_flap, y_bottom, "cut"))         # bottom edge

    # Bottom dust flap body bottom edge
    lines.append(_hline(x_body_left, x_body_right, y_bottom, "cut"))

    # Score line between dust flaps and bottom panel
    lines.append(_hline(x_left_flap, x_right_flap, y_dust_top, "score"))

    # === BOTTOM PANEL ===
    # Score line at top of bottom panel
    lines.append(_hline(x_body_left, x_body_right, y_bottom_panel_top, "score"))

    # Left side flap for bottom panel
    lines.append(_vline(x_left_flap, y_bottom_panel_top, y_dust_top, "cut"))
    lines.append(_hline(x_left_flap, x_body_left, y_bottom_panel_top, "cut"))

    # Right side flap for bottom panel
    lines.append(_vline(x_right_flap, y_bottom_panel_top, y_dust_top, "cut"))
    lines.append(_hline(x_body_right, x_right_flap, y_bottom_panel_top, "cut"))

    # Score lines at flap folds
    lines.append(_vline(x_body_left, y_dust_top, y_bottom_panel_top, "score"))
    lines.append(_vline(x_body_right, y_dust_top, y_bottom_panel_top, "score"))

    # === FRONT PANEL ===
    # Score at top of front panel
    lines.append(_hline(x_body_left, x_body_right, y_front_panel_top, "score"))

    # Left side flap for front panel
    lines.append(_vline(x_left_flap, y_front_panel_top, y_bottom_panel_top, "cut"))
    lines.append(_hline(x_left_flap, x_body_left, y_front_panel_top, "cut"))
    lines.append(_hline(x_left_flap, x_body_left, y_bottom_panel_top, "cut"))

    # Right side flap for front panel
    lines.append(_vline(x_right_flap, y_front_panel_top, y_bottom_panel_top, "cut"))
    lines.append(_hline(x_body_right, x_right_flap, y_front_panel_top, "cut"))
    lines.append(_hline(x_body_right, x_right_flap, y_bottom_panel_top, "cut"))

    # Score lines at flap folds
    lines.append(_vline(x_body_left, y_bottom_panel_top, y_front_panel_top, "score"))
    lines.append(_vline(x_body_right, y_bottom_panel_top, y_front_panel_top, "score"))

    # === TOP PANEL ===
    # Score at top of top panel
    lines.append(_hline(x_body_left, x_body_right, y_top_panel_top, "score"))

    # Left side flap for top panel
    lines.append(_vline(x_left_flap, y_top_panel_top, y_front_panel_top, "cut"))
    lines.append(_hline(x_left_flap, x_body_left, y_top_panel_top, "cut"))
    lines.append(_hline(x_left_flap, x_body_left, y_front_panel_top, "cut"))

    # Right side flap for top panel
    lines.append(_vline(x_right_flap, y_top_panel_top, y_front_panel_top, "cut"))
    lines.append(_hline(x_body_right, x_right_flap, y_top_panel_top, "cut"))
    lines.append(_hline(x_body_right, x_right_flap, y_front_panel_top, "cut"))

    # Score lines at flap folds
    lines.append(_vline(x_body_left, y_front_panel_top, y_top_panel_top, "score"))
    lines.append(_vline(x_body_right, y_front_panel_top, y_top_panel_top, "score"))

    # === TUCK FLAP ===
    # Tuck flap is slightly narrower than the body for easier insertion
    tuck_inset = t  # inset by material thickness on each side
    x_tuck_left = x_body_left + tuck_inset
    x_tuck_right = x_body_right - tuck_inset

    lines.append(_vline(x_tuck_left, y_top_panel_top, y_tuck_top, "cut"))    # left edge
    lines.append(_hline(x_tuck_left, x_tuck_right, y_tuck_top, "cut"))       # top edge
    lines.append(_vline(x_tuck_right, y_tuck_top, y_top_panel_top, "cut"))   # right edge

    # Angled cuts connecting tuck flap to body at top panel score line
    lines.append(Line(x1=x_body_left, y1=y_top_panel_top, x2=x_tuck_left, y2=y_top_panel_top, kind="cut"))
    lines.append(Line(x1=x_body_right, y1=y_top_panel_top, x2=x_tuck_right, y2=y_top_panel_top, kind="cut"))

    # Kerf compensation: expand the dieline uniformly so the laser kerf
    # (material removed by the cut) doesn't shrink final dimensions.
    k = kerf / 2
    if k > 0:
        for line in lines:
            line.x1 += k
            line.y1 += k
            line.x2 += k
            line.y2 += k
        total_w += kerf
        total_h += kerf

    return Dieline(width=total_w, height=total_h, lines=lines)

File: web/src/core/generators/mailer.ts
import type { BoxRequest, Dieline, Element } from "../models";
import { effectiveThickness, effectiveKerf, resolveEmbellishments } from "../models";
import { hline, vline, pathFromPoints, roundedPath, applyKerf } from "./helpers";
import {
  MAILER_TUCK_DEPTH_RATIO,
  MAILER_TUCK_TAPER_RATIO,
  MAILER_FLAP_WIDTH_RATIO,
  MAILER_LID_THICKNESS_RATIO,
  MAILER_BEVEL_INSET_MULTIPLIER,
} from "./proportions";

/**
 * Emit one side flap as a continuous path using roundedPath:
 *   score line → bevel → flap top → outer edge → flap bot → bevel → score line
 *
 * Uses roundedPath so adjacent corners (bevel + flap) correctly share edges
 * and radii are clamped to prevent overlap.
 */
function sideFlap(
  bodyTop: number, bodyBot: number,
  bevelTop: number, bevelBot: number,
  flapOuter: number,
  ft: number, fb: number,
  yScoreTop: number, yScoreBot: number,
  flapR: number, bevelR: number,
  kind: "cut" | "score",
): Element[] {
  return roundedPath(
    [
      [bodyTop, yScoreTop],    // 0: start at score line
      [bevelTop, ft],          // 1: bevel corner (in)
      [flapOuter, ft],         // 2: flap top-outer corner
      [flapOuter, fb],         // 3: flap bot-outer corner
      [bevelBot, fb],          // 4: bevel corner (out)
      [bodyBot, yScoreBot],    // 5: end at score line
    ],
    [bevelR, flapR, flapR, bevelR],
    kind,
  );
}

export function generateMailerDieline(request: BoxRequest): Dieline {
  const L = request.length;
  const W = request.width;
  const H = request.height;
  const T = effectiveThickness(request);
  const kerf = effectiveKerf(request);
  const emb = resolveEmbellishments(request);

  // Panel heights
  const tuckH = H * MAILER_TUCK_DEPTH_RATIO;
  const lidAdj = T * MAILER_LID_THICKNESS_RATIO;
  const topPanelH = request.lid === "inside" ? W - lidAdj : W + lidAdj;
  const frontPanelH = H;
  const bottomPanelH = W;
  const backPanelH = H;

  // Panel body widths
  const narrowBody = L;
  const wideBody = L + T;

  // Side flap widths
  const heightFlapW = W * MAILER_FLAP_WIDTH_RATIO;
  const widthFlapW = H * MAILER_FLAP_WIDTH_RATIO;

  // Bevel inset
  const bevel = T * MAILER_BEVEL_INSET_MULTIPLIER;

  // Total dieline
  let totalH = tuckH + topPanelH + frontPanelH + bottomPanelH + backPanelH;
  let totalW = Math.max(narrowBody + 2 * heightFlapW, wideBody + 2 * widthFlapW);

  const cx = totalW / 2;
  const narrowLeft = cx - narrowBody / 2;
  const narrowRight = cx + narrowBody / 2;
  const wideLeft = cx - wideBody / 2;
  const wideRight = cx + wideBody / 2;

  // Y coordinates
  const yTop = tuckH;
  const yFront = yTop + topPanelH;
  const yBottom = yFront + frontPanelH;
  const yBack = yBottom + bottomPanelH;
  const yEnd = totalH;

  // Flap edges (inset T from panel boundaries)
  const topFt = yTop + T;
  const topFb = yFront - T;
  const frontFt = yFront + T;
  const frontFb = yBottom - T;
  const bottomFt = yBottom + T;
  const bottomFb = yBack - T;
  const backFt = yBack + T;
  const backFb = yEnd - T;

  // Flap outer x positions
  const topFlapRight = wideRight + widthFlapW;
  const frontFlapRight = narrowRight + heightFlapW;
  const bottomFlapRight = wideRight + widthFlapW;
  const backFlapRight = narrowRight + heightFlapW;
  const topFlapLeft = wideLeft - widthFlapW;
  const frontFlapLeft = narrowLeft - heightFlapW;
  const bottomFlapLeft = wideLeft - widthFlapW;
  const backFlapLeft = narrowLeft - heightFlapW;

  // Tuck flap geometry
  const tuckBaseInset = T;
  const tuckTaper = H * MAILER_TUCK_TAPER_RATIO;
  const tuckBl = narrowLeft + tuckBaseInset;
  const tuckBr = narrowRight - tuckBaseInset;
  const tuckTl = tuckBl + tuckTaper;
  const tuckTr = tuckBr - tuckTaper;

  // Embellishment radii
  const flapR = emb.flapCornerRadius;
  const bevelR = emb.bevelRadius;
  const tuckR = emb.tuckCornerRadius;

  const elements: Element[] = [];

  // === CUT OUTLINE ===

  // --- TUCK (rounded at tip corners, sharp at base) ---
  elements.push(...roundedPath([
    [tuckTl, 0],
    [tuckTr, 0],
    [tuckBr, yTop],
    [narrowRight, yTop],
    [wideRight, yTop],
  ], [tuckR, 0, 0], "cut"));

  // --- RIGHT SIDE (going down) ---
  // Top panel flap
  elements.push(...sideFlap(
    wideRight, wideRight,
    wideRight + bevel, wideRight + bevel,
    topFlapRight, topFt, topFb, yTop, yFront,
    flapR, bevelR, "cut",
  ));

  // Transition: wide → narrow
  elements.push(...pathFromPoints([
    [wideRight, yFront], [narrowRight, yFront],
  ], "cut"));

  // Front panel flap
  elements.push(...sideFlap(
    narrowRight, narrowRight,
    narrowRight + bevel, narrowRight + bevel,
    frontFlapRight, frontFt, frontFb, yFront, yBottom,
    flapR, bevelR, "cut",
  ));

  // Transition: narrow → wide
  elements.push(...pathFromPoints([
    [narrowRight, yBottom], [wideRight, yBottom],
  ], "cut"));

  // Bottom panel flap
  elements.push(...sideFlap(
    wideRight, wideRight,
    wideRight + bevel, wideRight + bevel,
    bottomFlapRight, bottomFt, bottomFb, yBottom, yBack,
    flapR, bevelR, "cut",
  ));

  // Transition: wide → narrow
  elements.push(...pathFromPoints([
    [wideRight, yBack], [narrowRight, yBack],
  ], "cut"));

  // Back panel flap
  elements.push(...sideFlap(
    narrowRight, narrowRight,
    narrowRight + bevel, narrowRight + bevel,
    backFlapRight, backFt, backFb, yBack, yEnd,
    flapR, bevelR, "cut",
  ));

  // --- BOTTOM EDGE ---
  elements.push(hline(narrowRight, narrowLeft, yEnd, "cut"));

  // --- LEFT SIDE (going up, mirrored) ---
  // Back panel flap (left)
  elements.push(...sideFlap(
    narrowLeft, narrowLeft,
    narrowLeft - bevel, narrowLeft - bevel,
    backFlapLeft, backFb, backFt, yEnd, yBack,
    flapR, bevelR, "cut",
  ));

  // Transition: narrow → wide
  elements.push(...pathFromPoints([
    [narrowLeft, yBack], [wideLeft, yBack],
  ], "cut"));

  // Bottom panel flap (left)
  elements.push(...sideFlap(
    wideLeft, wideLeft,
    wideLeft - bevel, wideLeft - bevel,
    bottomFlapLeft, bottomFb, bottomFt, yBack, yBottom,
    flapR, bevelR, "cut",
  ));

  // Transition: wide → narrow
  elements.push(...pathFromPoints([
    [wideLeft, yBottom], [narrowLeft, yBottom],
  ], "cut"));

  // Front panel flap (left)
  elements.push(...sideFlap(
    narrowLeft, narrowLeft,
    narrowLeft - bevel, narrowLeft - bevel,
    frontFlapLeft, frontFb, frontFt, yBottom, yFront,
    flapR, bevelR, "cut",
  ));

  // Transition: narrow → wide
  elements.push(...pathFromPoints([
    [narrowLeft, yFront], [wideLeft, yFront],
  ], "cut"));

  // Top panel flap (left)
  elements.push(...sideFlap(
    wideLeft, wideLeft,
    wideLeft - bevel, wideLeft - bevel,
    topFlapLeft, topFb, topFt, yFront, yTop,
    flapR, bevelR, "cut",
  ));

  // --- Close back to tuck ---
  elements.push(...roundedPath([
    [wideLeft, yTop],
    [narrowLeft, yTop],
    [tuckBl, yTop],
    [tuckTl, 0],
  ], [0, 0], "cut"));

  // === SCORE LINES ===
  elements.push(hline(wideLeft, wideRight, yTop, "score"));
  elements.push(hline(wideLeft, wideRight, yFront, "score"));
  elements.push(hline(narrowLeft, narrowRight, yBottom, "score"));
  elements.push(hline(wideLeft, wideRight, yBack, "score"));

  const scoreRanges: [number, number, number, number][] = [
    [wideLeft, wideRight, yTop, yFront],
    [narrowLeft, narrowRight, yFront, yBottom],
    [wideLeft, wideRight, yBottom, yBack],
    [narrowLeft, narrowRight, yBack, yEnd],
  ];
  for (const [bl, br, yt, yb] of scoreRanges) {
    elements.push(vline(bl, yt, yb, "score"));
    elements.push(vline(br, yt, yb, "score"));
  }

  // Kerf compensation
  applyKerf(elements, kerf);
  if (kerf > 0) {
    totalW += kerf;
    totalH += kerf;
  }

  return { width: totalW, height: totalH, elements };
}
